# Data preparation


## Searching for the MAGs
1) Choose the MAGs   
For this project, the following bacterial species were selected: _Lactococcus lactis_, _Hafnia paralvei_, _Enterococcus faecalis_, _Bacteroides uniformis_, _Phocaeicola vulgatus_, _Parabacteroides goldsteinii_, _Citrobacter braaki_, _Akkermansia municiphila_, _Enterococcus hirae_ and _Bacteroides fragilis_.

EHI MAGs:
In the EHI database, select the MAGs with > 90% completeness and < 2.5 contamination for each of the species.

GTDB MAGs:
("NCBI Taxonomy" CONTAINS "Lactococcus lactis" AND "CheckM2 Completeness" > "90" AND "CheckM Contamination" < "2.5" AND ("Isolation Source" CONTAINS "feces" OR "Isolation Source" CONTAINS "excrement" OR "Isolation Source" CONTAINS "metagenome" OR "Isolation Source" CONTAINS "microbiome" OR "Isolation Source" CONTAINS "gut" OR "Isolation Source" CONTAINS "faeces" OR "Isolation Source" CONTAINS "fecal") AND "Isolation Source" IS NOT "N/A")

NCBI MAGs:
Refer to 03_downloading_mags.Rmd

## Downloading the MAGs and generating data
2) Download genome indices and metadata    
Download the EHI_MAG index for each species (in /data/mags_metadata folder) and the curl file and search metadata tsv from the GTDB and place in each species directory in Mjolnir.

2.5) Extract genome metadata   
- Use the GTDB search tsv to run this script to obtain more metadata. 

```{bash, eval = FALSE}
snakemake -s /maps/projects/alberdilab/people/pjq449/comparative_metagenomics/snakefiles/gtdb_metadata_pipeline.smk -j 1 --use-conda --rerun-incomplete
```

METADATA FILES:
-The EHI metadata files : data/mags_metadata/lactococcus_lactis_metadata.tsv
-The GTDB metadata files: data/mags_metadata/lactococcus_lactis_gtdb_final_metadata.tsv
-The NCBI metadata files: lactococcus_lactis_ncbi_metadata.rds 

3) Create the master index   
Run the create_index.R script to create an index of all the MAGs and the download paths (needs a list of species as input  and right now it is hardcoded in the script)
```{bash, eval = FALSE}
conda activate r_env
Rscript scripts/create_index.R 
```

4) Run the downloading_mags.smk to download all the genomes
```{bash, eval = FALSE}
module load snakemake
#testing

snakemake -s snakefiles/downloading_and_unzipping.smk \
  --executor slurm \
  --jobs 50 \
  --rerun-incomplete \
  --keep-going \
  --rerun-triggers mtime

```

Download the MAGs from the NCBI
```{bash, eval = FALSE}
### DOWNLOAD THE MAGS - add this to download and unzip mags??
conda activate drakkar_env 

#Run this inside the mags folder
datasets download genome accession --inputfile ../phocaeicola_vulgatus_ncbi_selected_accessions.txt --include genome --filename phocaeicola_vulgatus_ncbi_selected_genomes.zip

#then unzip
zip="phocaeicola_vulgatus_ncbi_selected_genomes.zip"

mkdir -p unzipped

unzip -Z1 "$zip" | grep '\.fna$' | while read -r f; do
    acc=$(basename "$(dirname "$f")")
    echo "Extracting $acc"
    unzip -p "$zip" "$f" > "unzipped/${acc}.fna"
done

```


5) Run drep
```{bash, eval = FALSE}
sbatch scripts/drep_compare.slurm
```

6) Remove the MAGs that are <95% ANI (not the same species)
```{bash, eval = FALSE}
python3 scripts/filter_by_cdb.py data data/clusters_to_drop.tsv 
```

6.2) Re-run drep without removed MAGs
```{bash, eval = FALSE}
sbatch --export=SPECIES="$SPECIES" scripts/rerun_drep_removed.sh
```


6) Make a screen session for each species.
```{bash, eval = FALSE}
screen -S parabacteroides_distasonis
```


7) Run drakkar annotating_function.smk to re-annotate all the MAGs:
```{bash, eval = FALSE}
drakkar annotating -b /maps/projects/alberdilab/people/pjq449/comparative_metagenomics/data/phocaeicola_vulgatus/mags/unzipped -o /maps/projects/alberdilab/people/pjq449/comparative_metagenomics/data/phocaeicola_vulgatus --env_path /projects/alberdilab/data/environments/drakkar --annotation-type function 
```


7.1) Run contig to genome mapping
```{bash, eval = FALSE}
bash /maps/projects/alberdilab/people/pjq449/comparative_metagenomics/scripts/build_contig_to_mag_maps_all.sh  # though in the end i just ran one in terminal directly
```


8) Pangenome analysis with ppanggolin
```{bash, eval = FALSE}
snakemake --snakefile Snakefile.pangolin \
          --configfile snakefiles/pangolin_config.yaml \
          --use-conda \
          -j 40
```

