
## Data preparation

### EHI MAGs
```{r load ehi mags first_human}
ehi_mags <- read_csv("data/ehi_mags.csv")
```
### Prepare color scheme

[AlberdiLab](www.alberdilab.dk) projects use unified color schemes developed for the [Earth Hologenome Initiative](www.earthhologenome.org), to facilitate figure interpretation.

```{r get_ehi_colors_human, warning=FALSE, comments="", message=FALSE, eval=FALSE}
ehi_mags_p <- ehi_mags %>%
  mutate(phylum=str_remove_all(phylum, "p__")) 

phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    mutate(phylum=str_remove_all(phylum, "p__")) %>%
    right_join(ehi_mags_p, by=join_by(phylum == phylum)) %>%
    dplyr::select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)

```


```{r colors_human}
#source_colors <- c(EHI = "#8BC63F", GTDB = "#2D522D")
source_colors_human <- c(EHI = "#8BC63F", GTDB = "#A90D00")
```

### Sample metadata
```{r load gtdb search metadata}
gtdb_search_metadata <- read_tsv("data/human/gtdb_search_results.tsv") 



colnames(gtdb_search_metadata)
```

```{r load gtdb accession metadata_human}
gtdb_accession_metadata<- read_tsv("data/human/parabacteroides_distasonis_human_GTDB_accession_metadata.tsv") 
```


```{r load EHI metadata_human}
ehi_metadata <- read_tsv("data/mags_metadata/parabacteroides_distasonis_metadata.tsv") %>%
  mutate(GC = as.numeric(str_remove(GC, "%")))
```
### Load genome mapping


```{r join the metadatas_human}
# Prepare EHI metadata with standardized column names
ehi_clean <- ehi_metadata %>%
  dplyr::select(
    ID,
    species,  
    completeness,
    contamination,
    genome_size = size,
    GC,
    N50,
    contigs,
    host_species,
    host_order,
    host_class,
    assembly_type,
    isolation_source= sample_type,
    mag_name, 
    eha_number
  ) %>%
  mutate(source = "EHI")

# Prepare GTDB metadata
# Join search and accession metadata
gtdb_metadata <- full_join(gtdb_accession_metadata, gtdb_search_metadata,  by = c("search_accession" = "Accession")
  )



# Prepare GTDB metadata
gtdb_clean <- gtdb_metadata %>%
  dplyr::select(
    ID = search_accession,
    completeness = "checkm2_completeness",
    contamination = "checkm2_contamination",
    gtdb_taxonomy = "GTDB Taxonomy",
    isolation_source = "Isolation Source",
    genome_size,
    GC = gc_percentage,
    N50 = n50_contigs,
    contigs = contig_count,
    ncbi_country,
    mimag_high_quality,
    mimag_medium_quality,
    gtdb_representative,
    mag_name = search_accession
  )%>%
  mutate(source = "GTDB")

#Join
genome_metadata <- bind_rows(ehi_clean, gtdb_clean)

```




```{r load master index_human}
# Read master index
master_index <- read_tsv("data/human/master_mag_index_2.tsv") %>%
  filter(species == "parabacteroides_distasonis") %>%
  mutate(
    # Extract the actual genome identifier from out_path
    genome_identifier = case_when(
      # For EHI: extract EHA00531_bin.1 from the filename
      str_detect(out_path, "EHA") ~ str_extract(out_path, "EHA[0-9]+_bin\\.[0-9]+"),
      # For NCBI: use the ID as is (GCA/GCF number)
      TRUE ~ ID
    )
  )
```

```{r mapping the ids to the genome identifiers_human}
# Check the mapping
print("Master index mapping:")
master_index %>% 
  dplyr::select(ID, genome_identifier, out_path) %>%
  head(10) %>%
  print()

# Read contig-to-genome mapping
contig_to_genome <- read_tsv("data/human/contig_to_mag.tsv",
                              col_names = c("contig", "genome_filename"))
```


### Genome annotations

```{r e_human}
genome_annotations <-read_tsv("data/human/gene_annotations_human.tsv.xz")
```


```{r load genome annotations with ids_human}
# Read annotations and add IDs
genome_annotations <- read_tsv("data/human/gene_annotations_human.tsv.xz") %>%
  mutate(contig = sub("_[^_]*$", "", gene)) %>%
  left_join(contig_to_genome, by = "contig") %>%
  mutate(genome= genome_filename)%>%
   filter(!is.na(genome))
```

### Distill annotations into GIFTs 

```{r distill_annotations_human, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol= 15, annotcol=c(5,6,7,8), verbosity = T)
```



### Load trees
```{r load trees_human}
genome_metadata$mag_name <- sub("\\.fa$", "", genome_metadata$mag_name)#remove .fa from the mag names so that they match the tree ids


getphylo_tree <- read_tree("data/combined_alignment.tree")
getphylo_tree$tip.label <- gsub("'", "", getphylo_tree$tip.label)
tip_df <- data.frame(label = getphylo_tree$tip.label) %>%
  left_join(genome_metadata, by = c("label" = "mag_name"))

#replace long mag names with IDs
getphylo_tree$tip.label <- tip_df$ID

```


## Wrap working objects

All working objects are wrapped into a single Rdata object to facilitate downstream usage.

```{r wrap_working_objects_human, warning=FALSE, comments="", message=FALSE, eval=FALSE}
save(ehi_mags,
     phylum_colors,
     genome_annotations,
     genome_gifts,
     contig_to_genome,
     gtdb_metadata,
     ehi_metadata,
     master_index,
     genome_metadata,
     source_colors,
     getphylo_tree,
     file = "data/human/data.Rdata")
```

